### PR TITLE
feat: add Client.logout and context manager

### PR DESCRIPTION
**Pull request by oleobal**
## Summary

Add a `Client.logout` function, which deletes a token obtained by `Client.login` (if any).

Add a context manager, making the following snippet legal:

```python
with Client(
    client_name="org-1",
    backend_type="remote",
    url="http://substra-backend.org-1.com:8000",
    username="org-1",
    password="p@sswr0d44",
) as client:
    pass
```

Requires https://github.com/Substra/substra-backend/pull/698

Closes FL-1008

## Please check if the PR fulfills these requirements

- [x] If necessary, the [changelog](https://github.com/Substra/substra/blob/main/CHANGELOG.md) has been updated
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] The commit message follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification
- For any breaking changes, companion PRs have been opened on the following repositories:
  - [ ] [substra-tests](https://github.com/Substra/substra)
  - [ ] [substrafl](https://github.com/Substra/substrafl)
  - [ ] [substra-documentation](https://github.com/Substra/substra-documentation)

### CODE REVIEW

File: substra/sdk/client.py
Comment (oleobal) [inline substra/sdk/client.py:321]: Unsure whether context manager is still needed with `__del__` defined, but people online say not to rely on it and using `with` is preferable. However that does mean updating our documentation.

File: substra/sdk/client.py
Comment (ThibaultFy) [inline substra/sdk/client.py:390]: I would check the client type rather than `hasattr` if the purpose of this is to check that we have a remote client.

File: substra/sdk/client.py
Comment (ThibaultFy) [reply substra/sdk/client.py:321]: What part of the documentation are you reffering to ?

**Review by ThibaultFy — COMMENTED** [review]
Thanks for the PR :) 

Concerning the fact that we define a logout function only for the remote backend, I am not 100% convince... There is a login method defined in locals backend too, I think for the hybrid mode. You should take a look at it to ensure what we should do for this mode I think. (cf `sdk/backends/local/dal.py`.

For the tests, you could check that the token is deleted the using the context manager `with` or when deleting the object (`del client`).
This test can be useful to mock minimally https://github.com/Substra/substra/blob/eae5d52ab804c9a4d6794ef299a078790069b253/tests/sdk/test_client.py#L92

File: substra/sdk/client.py
Comment (oleobal) [reply substra/sdk/client.py:321]: I mean that in the documentation all examples of the form:

```
client = Client(...)
do_stuff(client)
```

should be replaced with:


```python
with Client(...) as client:
    do_stuff(client)
```

File: substra/sdk/client.py
Comment (oleobal) [reply substra/sdk/client.py:390]: That makes sense, but isn't the purpose at this level just to check whether the client has a `logout` function, regardless of what kind of client it is?

Comment (ThibaultFy) [reply substra/sdk/client.py:390]: Regarding my main comment, I think it would make more sense that all client have a logout, as they all have a login method

File: substra/sdk/client.py
Comment (ThibaultFy) [reply substra/sdk/client.py:321]: Some scripts (such as the example) need the client for more than a thousand lines, no I would not replace the old version, maybe just indicate that the context manager is also a valid option ?

Comment (oleobal) [reply substra/sdk/client.py:327]: I added it in the `sdk.Client` reference, under the `username` argument.

File: substra/sdk/client.py
Comment (ThibaultFy) [inline substra/sdk/client.py:203]:
```suggestion
            If using username/password, you should use a context manager to ensure the session terminates as intended:
```

**Review by ThibaultFy — COMMENTED** [review]
A few small comments and will be good to me ! do you think we need a run of CI on it before merging ? No strong opinion on it on my side.

**Comment by oleobal** [comment]
@ThibaultFy 
> do you think we need a run of CI on it before merging ? No strong opinion on it on my side.

I [ran one yesterday](https://github.com/Substra/substra-backend/pull/698#issuecomment-1680434880) but let's run a new one just to be sure ^^

**Comment by ThibaultFy** [comment]
> @ThibaultFy
> 
> > do you think we need a run of CI on it before merging ? No strong opinion on it on my side.
> 
> I [ran one yesterday](https://github.com/Substra/substra-backend/pull/698#issuecomment-1680434880) but let's run a new one just to be sure ^^

Ah ok ok :) you're not using the /e2e command ?

**Comment by oleobal** [comment]
> you're not using the /e2e command ?

I am, just from the backend PR instead of this one

**Review by ThibaultFy — APPROVED** [review]
LGTM, thanks a lot :)

**Pull request merged**
